File: android_native/docx/src/main/kotlin/com/afluffywaffle/layuv/docx/RewriteProtocol.kt
package com.afluffywaffle.layuv.docx

/**
 * Wire protocol for the in-app "Ask AI" conversation. The chat is a back-and-forth:
 * the model may discuss or ask clarifying questions in plain text, and when it
 * delivers a full chapter rewrite it wraps ONLY the manuscript prose between the
 * [BEGIN] and [END] marker lines. The app renders discussion inline but turns a
 * rewrite into a "save as a new draft" action — it never dumps the chapter into the
 * chat. So this is the single source of truth for BOTH the prompt instructions
 * ([ManuscriptSerializer]) and the reply parser (the reader's Ask AI panel).
 */
object RewriteProtocol {
    const val BEGIN = "===REWRITE==="
    const val END = "===END REWRITE==="

    /** [conversation] = discussion text (may be empty); [rewrite] = chapter prose, or null if none. */
    data class Parsed(val conversation: String, val rewrite: String?)

    /**
     * Split a reply into its discussion text and (if present) the rewritten chapter.
     * A reply with no [BEGIN] marker is pure conversation. An unterminated block
     * (truncated mid-rewrite) treats everything after [BEGIN] as the partial rewrite.
     */
    fun parse(text: String): Parsed {
        val begin = text.indexOf(BEGIN)
        if (begin < 0) return Parsed(text.trim(), null)
        val conversation = text.substring(0, begin).trim()
        val afterBegin = begin + BEGIN.length
        val end = text.indexOf(END, afterBegin)
        val rewrite = (if (end < 0) text.substring(afterBegin) else text.substring(afterBegin, end)).trim()
        return Parsed(conversation, rewrite.ifEmpty { null })
    }
}

File: macos_native/Packages/LeamhDocx/Sources/LeamhDocx/RewriteProtocol.swift
import Foundation

/// Wire protocol for the in-app "Ask AI" conversation. The model discusses or asks
/// questions in plain text, and when it delivers a full chapter rewrite it wraps ONLY
/// the manuscript prose between the `begin` and `end` marker lines. The app renders
/// discussion inline but turns a rewrite into a "save as a new draft" action. Single
/// source of truth for both the prompt instructions (ManuscriptSerializer) and the
/// reply parser (the Ask AI panel). Mirror of RewriteProtocol.kt.
public enum RewriteProtocol {
    public static let begin = "===REWRITE==="
    public static let end = "===END REWRITE==="

    /// `conversation` = discussion text (may be empty); `rewrite` = chapter prose, or nil if none.
    public struct Parsed: Equatable {
        public let conversation: String
        public let rewrite: String?
        public init(conversation: String, rewrite: String?) {
            self.conversation = conversation
            self.rewrite = rewrite
        }
    }

    /// Split a reply into its discussion text and (if present) the rewritten chapter.
    /// A reply with no `begin` marker is pure conversation. An unterminated block
    /// (truncated mid-rewrite) treats everything after `begin` as the partial rewrite.
    public static func parse(_ text: String) -> Parsed {
        let ws = CharacterSet.whitespacesAndNewlines
        guard let beginRange = text.range(of: begin) else {
            return Parsed(conversation: text.trimmingCharacters(in: ws), rewrite: nil)
        }
        let conversation = String(text[..<beginRange.lowerBound]).trimmingCharacters(in: ws)
        let afterBegin = beginRange.upperBound
        let rewriteRaw: String
        if let endRange = text.range(of: end, range: afterBegin..<text.endIndex) {
            rewriteRaw = String(text[afterBegin..<endRange.lowerBound])
        } else {
            rewriteRaw = String(text[afterBegin...])
        }
        let rewrite = rewriteRaw.trimmingCharacters(in: ws)
        return Parsed(conversation: conversation, rewrite: rewrite.isEmpty ? nil : rewrite)
    }
}
